driver/src/test/scala/za/co/absa/hyperdrive/driver/drivers/KafkaToKafkaDeduplicationAfterRetryDockerTest.scala

```diff
@@ -0,0 +1,246 @@
+    val config = new Properties()
+    config.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaSchemaRegistryWrapper.kafka.getBootstrapServers)
+    val localKafkaAdmin = AdminClient.create(config)
+    val replication = 1.toShort
+    val topic = new NewTopic(topicName, partitions, replication)
+    val topicCreationFut = localKafkaAdmin.createTopics(util.Arrays.asList(topic)).all()
+    while(!topicCreationFut.isDone) {}
+  }
+
+  it should "execute the whole kafka-to-kafka pipeline" in {
+    // given
+    val kafkaSchemaRegistryWrapper = new KafkaSchemaRegistryWrapper
+    val sourceTopic = "deduplication_src"
+    val destinationTopic = "deduplication_dest"
+    val sourceTopicPartitions = 5
+    val destinationTopicPartitions = 3
+    val schemaManager = SchemaManagerFactory.create(Map("schema.registry.url" -> kafkaSchemaRegistryWrapper.schemaRegistryUrl))
+    val subject = SchemaSubject.usingTopicNameStrategy(sourceTopic)
+    val parserV1 = new Parser()
+    val schemaV1 = parserV1.parse(schemaV1String(sourceTopic))
+    val parserV2 = new Parser()
+    val schemaV2 = parserV2.parse(schemaV2String(sourceTopic))
+    val schemaV1Id = schemaManager.register(subject, schemaV1)
+    val schemaV2Id = schemaManager.register(subject, schemaV2)
+
+    val producer = createProducer(kafkaSchemaRegistryWrapper)
+    createTopic(kafkaSchemaRegistryWrapper, sourceTopic, sourceTopicPartitions)
+    createTopic(kafkaSchemaRegistryWrapper, destinationTopic, destinationTopicPartitions)
+
+    val recordsV1 = (0 until 50).map(i => {
+      val valueRecord = new GenericData.Record(schemaV1)
+      valueRecord.put("record_id", i)
+      valueRecord.put("value_field", s"valueHello_$i")
+      valueRecord
+    })
+    val recordsV2 = (50 until 100).map(i => {
+      val valueRecord = new GenericData.Record(schemaV2)
+      valueRecord.put("record_id", i)
+      valueRecord.put("value_field", null)
+      valueRecord
+    })
+    sendData(producer, recordsV1, sourceTopic, sourceTopicPartitions)
+    sendData(producer, recordsV2, sourceTopic, sourceTopicPartitions)
+
+    Thread.sleep(3000)
+
+    val driverConfig = Map(
+      // Pipeline settings
+      "component.ingestor" -> "spark",
+      "component.reader" -> "za.co.absa.hyperdrive.ingestor.implementation.reader.kafka.KafkaStreamReader",
+      "component.transformer.id.0" -> "[column.copy]",
+      "component.transformer.class.[column.copy]" -> "za.co.absa.hyperdrive.ingestor.implementation.transformer.column.copy.ColumnCopyStreamTransformer",
+      "component.transformer.id.1" -> "[avro.decoder]",
+      "component.transformer.class.[avro.decoder]" -> "za.co.absa.hyperdrive.ingestor.implementation.transformer.avro.confluent.ConfluentAvroDecodingTransformer",
+      "component.transformer.id.2" -> "[kafka.deduplicator]",
+      "component.transformer.class.[kafka.deduplicator]" -> "za.co.absa.hyperdrive.ingestor.implementation.transformer.deduplicate.kafka.DeduplicateKafkaSinkTransformer",
```

Comment:
Can you run this test with a different configuration where you don't have DeduplicateKafkaSinkTransformer ?
You can expect duplicates in sink and it is predictable which message should be duplicated.

Reply:
Yes I can do that